panic in the TestManagerDispatchLoop

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

```
 [2024/01/15 03:03:17.975 +00:00] [ERROR] [spool.go:139] ["recover panic"] [recover="runtime error: index out of range [2227] with length 6"] [stack="github.com/pingcap/tidb/pkg/resourcemanager/pool/spool.(*Pool).run.func1.1\n\tpkg/resourcemanager/pool/spool/spool.go:139\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:914\nruntime.goPanicIndex\n\tGOROOT/src/runtime/panic.go:114\ngithub.com/pingcap/tidb/pkg/disttask/framework/scheduler_test.TestManagerDispatchLoop.func3.1\n\tpkg/disttask/framework/scheduler/scheduler_test.go:550\nreflect.Value.call\n\tGOROOT/src/reflect/value.go:596\nreflect.Value.Call\n\tGOROOT/src/reflect/value.go:380\ngo.uber.org/mock/gomock.(*Call).Do.func1\n\texternal/org_uber_go_mock/gomock/call.go:179\ngo.uber.org/mock/gomock.(*Controller).Call\n\texternal/org_uber_go_mock/gomock/controller.go:229\ngithub.com/pingcap/tidb/pkg/disttask/framework/mock.(*MockScheduler).ScheduleTask\n\tpkg/disttask/framework/mock/scheduler_mock.go:171\ngithub.com/pingcap/tidb/pkg/disttask/framework/scheduler.(*Manager).startScheduler.func1\n\tpkg/disttask/framework/scheduler/scheduler_manager.go:321\ngithub.com/pingcap/tidb/pkg/resourcemanager/pool/spool.(*Pool).run.func1\n\tpkg/resourcemanager/pool/spool/spool.go:144"] [stack="github.com/pingcap/tidb/pkg/resourcemanager/pool/spool.(*Pool).run.func1.1\n\tpkg/resourcemanager/pool/spool/spool.go:139\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:914\nruntime.goPanicIndex\n\tGOROOT/src/runtime/panic.go:114\ngithub.com/pingcap/tidb/pkg/disttask/framework/scheduler_test.TestManagerDispatchLoop.func3.1\n\tpkg/disttask/framework/scheduler/scheduler_test.go:550\nreflect.Value.call\n\tGOROOT/src/reflect/value.go:596\nreflect.Value.Call\n\tGOROOT/src/reflect/value.go:380\ngo.uber.org/mock/gomock.(*Call).Do.func1\n\texternal/org_uber_go_mock/gomock/call.go:179\ngo.uber.org/mock/gomock.(*Controller).Call\n\texternal/org_uber_go_mock/gomock/controller.go:229\ngithub.com/pingcap/tidb/pkg/disttask/framework/mock.(*MockScheduler).ScheduleTask\n\tpkg/disttask/framework/mock/scheduler_mock.go:171\ngithub.com/pingcap/tidb/pkg/disttask/framework/scheduler.(*Manager).startScheduler.func1\n\tpkg/disttask/framework/scheduler/scheduler_manager.go:321\ngithub.com/pingcap/tidb/pkg/resourcemanager/pool/spool.(*Pool).run.func1\n\tpkg/resourcemanager/pool/spool/spool.go:144"]
    scheduler_test.go:602: 
        	Error Trace:	pkg/disttask/framework/scheduler/scheduler_test.go:602
        	Error:      	Condition never satisfied
        	Test:       	TestManagerDispatchLoop 
```

https://tiprow.hawkingrei.com/view/gs/pingcapprow/pr-logs/pull/pingcap_tidb/50405/tiprow_fast_test/1746727973362339840#1:build-log.txt%3A4492

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

--- a/pkg/disttask/framework/scheduler/scheduler_test.go
+++ b/pkg/disttask/framework/scheduler/scheduler_test.go
@@ -496,7 +496,7 @@ func TestIsCancelledErr(t *testing.T) {
 	require.True(t, scheduler.IsCancelledErr(errors.New("cancelled by user")))
 }
 
-func TestManagerDispatchLoop(t *testing.T) {
+func TestManagerScheduleLoop(t *testing.T) {
 	// Mock 16 cpu node.
 	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/pkg/util/cpu/mockNumCpu", "return(16)"))
 	t.Cleanup(func() {

--- a/pkg/disttask/framework/scheduler/scheduler_test.go
+++ b/pkg/disttask/framework/scheduler/scheduler_test.go
@@ -527,6 +527,9 @@ func TestManagerScheduleLoop(t *testing.T) {
 	}
 	concurrencies := []int{4, 6, 16, 2, 4, 4}
 	waitChannels := make(map[string](chan struct{}))
+	for i := 0; i < len(concurrencies); i++ {
+		waitChannels[fmt.Sprintf("key/%d", i)] = make(chan struct{})
+	}
 	scheduler.RegisterSchedulerFactory(proto.TaskTypeExample,
 		func(ctx context.Context, task *proto.Task, param scheduler.Param) scheduler.Scheduler {
 			mockScheduler = mock.NewMockScheduler(ctrl)
@@ -550,10 +553,8 @@ func TestManagerScheduleLoop(t *testing.T) {
 						proto.TaskStateSucceed, proto.StepDone, task.ID)
 					return err
 				}))
-
 			})
 			mockScheduler.EXPECT().Close()
-			waitChannels[task.Key] = make(chan struct{})
 			return mockScheduler
 		},
 	)

--- a/pkg/disttask/framework/scheduler/scheduler_test.go
+++ b/pkg/disttask/framework/scheduler/scheduler_test.go
@@ -535,9 +535,6 @@ func TestManagerScheduleLoop(t *testing.T) {
 	scheduler.RegisterSchedulerFactory(proto.TaskTypeExample,
 		func(ctx context.Context, task *proto.Task, param scheduler.Param) scheduler.Scheduler {
 			idx := counter.Load()
-			if int(idx) > len(concurrencies) {
-				return nil
-			}
 			mockScheduler = mock.NewMockScheduler(ctrl)
 			// below 2 are for balancer loop, it's async, cannot determine how
 			// many times it will be called.
@@ -545,6 +542,9 @@ func TestManagerScheduleLoop(t *testing.T) {
 			mockScheduler.EXPECT().GetEligibleInstances(gomock.Any(), gomock.Any()).Return(nil, nil).AnyTimes()
 			mockScheduler.EXPECT().Init().Return(nil)
 			mockScheduler.EXPECT().ScheduleTask().Do(func() {
+				if task.IsDone() {
+					return
+				}
 				require.NoError(t, taskMgr.WithNewSession(func(se sessionctx.Context) error {
 					_, err := sqlexec.ExecSQL(ctx, se, "update mysql.tidb_global_task set state=%?, step=%? where id=%?",
 						proto.TaskStateRunning, proto.StepOne, task.ID)
@@ -556,9 +556,9 @@ func TestManagerScheduleLoop(t *testing.T) {
 						proto.TaskStateSucceed, proto.StepDone, task.ID)
 					return err
 				}))
+				counter.Add(1)
 			})
 			mockScheduler.EXPECT().Close()
-			counter.Add(1)
 			return mockScheduler
 		},
 	)

--- a/pkg/disttask/framework/scheduler/scheduler_test.go
+++ b/pkg/disttask/framework/scheduler/scheduler_test.go
@@ -535,6 +535,9 @@ func TestManagerDispatchLoop(t *testing.T) {
 	scheduler.RegisterSchedulerFactory(proto.TaskTypeExample,
 		func(ctx context.Context, task *proto.Task, param scheduler.Param) scheduler.Scheduler {
 			idx := counter.Load()
+			if int(idx) > len(concurrencies) {
+				return nil
+			}
 			mockScheduler = mock.NewMockScheduler(ctrl)
 			// below 2 are for balancer loop, it's async, cannot determine how
 			// many times it will be called.

--- a/pkg/disttask/framework/scheduler/scheduler_test.go
+++ b/pkg/disttask/framework/scheduler/scheduler_test.go
@@ -19,7 +19,6 @@ import (
 	"fmt"
 	"slices"
 	"strings"
-	"sync/atomic"
 	"testing"
 	"time"
 
@@ -527,14 +526,9 @@ func TestManagerScheduleLoop(t *testing.T) {
 		testutil.InsertSubtask(t, taskMgr, 1000000, proto.StepOne, execID, []byte(""), proto.SubtaskStatePending, proto.TaskTypeExample, 16)
 	}
 	concurrencies := []int{4, 6, 16, 2, 4, 4}
-	waitChannels := make([]chan struct{}, len(concurrencies))
-	for i := range waitChannels {
-		waitChannels[i] = make(chan struct{})
-	}
-	var counter atomic.Int32
+	waitChannels := make(map[string](chan struct{}))
 	scheduler.RegisterSchedulerFactory(proto.TaskTypeExample,
 		func(ctx context.Context, task *proto.Task, param scheduler.Param) scheduler.Scheduler {
-			idx := counter.Load()
 			mockScheduler = mock.NewMockScheduler(ctrl)
 			// below 2 are for balancer loop, it's async, cannot determine how
 			// many times it will be called.
@@ -550,15 +544,16 @@ func TestManagerScheduleLoop(t *testing.T) {
 						proto.TaskStateRunning, proto.StepOne, task.ID)
 					return err
 				}))
-				<-waitChannels[idx]
+				<-waitChannels[task.Key]
 				require.NoError(t, taskMgr.WithNewSession(func(se sessionctx.Context) error {
 					_, err := sqlexec.ExecSQL(ctx, se, "update mysql.tidb_global_task set state=%?, step=%? where id=%?",
 						proto.TaskStateSucceed, proto.StepDone, task.ID)
 					return err
 				}))
-				counter.Add(1)
+
 			})
 			mockScheduler.EXPECT().Close()
+			waitChannels[task.Key] = make(chan struct{})
 			return mockScheduler
 		},
 	)
@@ -583,7 +578,7 @@ func TestManagerScheduleLoop(t *testing.T) {
 			taskKeys[2] == "key/3" && taskKeys[3] == "key/4"
 	}, time.Second*10, time.Millisecond*100)
 	// finish the first task
-	close(waitChannels[0])
+	close(waitChannels["key/0"])
 	require.Eventually(t, func() bool {
 		taskKeys := getRunningTaskKeys()
 		return err == nil && len(taskKeys) == 4 &&
@@ -591,7 +586,7 @@ func TestManagerScheduleLoop(t *testing.T) {
 			taskKeys[2] == "key/4" && taskKeys[3] == "key/5"
 	}, time.Second*10, time.Millisecond*100)
 	// finish the second task
-	close(waitChannels[1])
+	close(waitChannels["key/1"])
 	require.Eventually(t, func() bool {
 		taskKeys := getRunningTaskKeys()
 		return err == nil && len(taskKeys) == 4 &&
@@ -600,7 +595,7 @@ func TestManagerScheduleLoop(t *testing.T) {
 	}, time.Second*10, time.Millisecond*100)
 	// close others
 	for i := 2; i < len(concurrencies); i++ {
-		close(waitChannels[i])
+		close(waitChannels[fmt.Sprintf("key/%d", i)])
 	}
 	require.Eventually(t, func() bool {
 		taskKeys := getRunningTaskKeys()

